### PR TITLE
add support for upgrading helm releases

### PR DESCRIPTION
**Pull request by sameersbn**
Implements Helm Release feature

Status:
 - User can deploy changes to an existing release. 
 - The values.yaml file is populated from the existing release, so the user will be presented the values that were used previously
 - Currently I have disabled the version selection which means only changes to values.yaml are possible. I need to figure out how to propogate the chartVersion changes to the Deployment form since this is a property field of the component.

### CODE REVIEW

File: dashboard/src/actions/charts.ts [hunk 1]
```diff
@@ -118,12 +118,17 @@ export function deployChart(
   releaseName: string,
   namespace: string,
   values?: string,
+  resourceVersion?: string,
```
Comment (sameersbn) [inline dashboard/src/actions/charts.ts]: `resourceVersion` is an optional parameter that signals that we're trying to upgrade an existing release

File: dashboard/src/components/DeploymentForm/DeploymentForm.tsx [hunk 2]
```diff
@@ -62,6 +65,15 @@ class DeploymentForm extends React.Component<IDeploymentFormProps, IDeploymentFo
...
     getChartVersion(chartID, chartVersion);
     getChartValues(chartID, chartVersion);
+
+    if (hr) {
+      this.setState({
+        appValues: hr.spec.values,
+        namespace: hr.metadata.namespace,
+        releaseName: hr.metadata.name,
+        valuesModified: true,
+      });
```
Comment (sameersbn) [inline dashboard/src/components/DeploymentForm/DeploymentForm.tsx]: if a release object is provided to the deployment form then the component state is initialized with parameters from the release object.

File: dashboard/src/components/DeploymentForm/DeploymentForm.tsx [hunk 3]
```diff
@@ -152,6 +165,7 @@ class DeploymentForm extends React.Component<IDeploymentFormProps, IDeploymentFo
                   onChange={this.handleChartVersionChange}
                   value={this.props.chartVersion}
                   required={true}
+                  disabled={hr ? true : false}
```
Comment (sameersbn) [inline dashboard/src/components/DeploymentForm/DeploymentForm.tsx]: at the moment I have disabled the version selection while performing an upgrade. I need to figure out how to propagate the version change in the form

File: dashboard/src/shared/HelmRelease.ts [hunk 4]
```diff
@@ -142,7 +142,7 @@ export class HelmRelease {
   private static parseRelease(hr: IHelmRelease, cm: IHelmReleaseConfigMap): IApp {
     const protoBytes = inflate(atob(cm.data.release));
     const rel = hapi.release.Release.decode(protoBytes);
-    const app: IApp = { data: rel, type: "helm" };
+    const app: IApp = { data: rel, type: "helm", hr };
```
Comment (sameersbn) [inline dashboard/src/shared/HelmRelease.ts]: adds the `IHelmRelease` to `IApp`

File: dashboard/src/actions/charts.ts [hunk 5]
```diff
@@ -118,12 +118,17 @@ export function deployChart(
...
 ) {
   return (dispatch: Dispatch<IStoreState>): Promise<{}> => {
     const chartAttrs = chartVersion.relationships.chart.data;
-    return fetch(url.api.helmreleases.create(namespace), {
+    const method = resourceVersion ? "PUT" : "POST";
```
Comment (prydonius) [inline dashboard/src/actions/charts.ts]: This is interesting, but it might be better to use a separate action for upgrade rather than using the same one to keep the intent clear. Also ideally we should move to the new way of making requests (see the Function class in `shared/Function.ts`), but we can do that in a separate PR.

File: dashboard/src/components/AppEdit/AppEdit.tsx [hunk 6]
```diff
@@ -0,0 +1,57 @@
+import * as React from "react";
+
+import { RouterAction } from "react-router-redux";
+import { IServiceBinding } from "../../shared/ServiceBinding";
+import { IApp, IChartState, IChartVersion } from "../../shared/types";
+
+import DeploymentForm from "../../components/DeploymentForm";
+
+interface IAppEditProps {
```
Comment (prydonius) [inline dashboard/src/components/AppEdit/AppEdit.tsx]: please cleanup unused props.

File: dashboard/src/shared/HelmRelease.ts (hunk 4 above)
Comment (prydonius) [reply dashboard/src/shared/HelmRelease.ts]: woot :+1:

File: dashboard/src/shared/url.ts [hunk 7]
```diff
@@ -42,6 +42,8 @@ export const api = {
       `/api/kube/api/v1/namespaces/kubeapps/configmaps?labelSelector=NAME in (${releaseNames.join(
         ",",
       )})`,
+    upgrade: (namespace = "default", releaseName: string) =>
```
Comment (prydonius) [inline dashboard/src/shared/url.ts]: Just FYI, we stopped using this and switched to classes to handle API requests, see the classes under `shared/` for examples. That said, more of an FYI as we can go back and update this in a future PR.

File: dashboard/src/components/DeploymentForm/DeploymentForm.tsx [hunk 8]
```diff
@@ -243,7 +256,15 @@ class DeploymentForm extends React.Component<IDeploymentFormProps, IDeploymentFo
     this.setState({ releaseName: e.currentTarget.value });
   };
   public handleChartVersionChange = (e: React.FormEvent<HTMLSelectElement>) => {
-    this.props.push(`/apps/new/${this.props.chartID}/versions/${e.currentTarget.value}`);
+    const { hr } = this.props;
+    let pushUrl;
+    if (hr) {
+      const { releaseName, namespace } = this.state;
+      pushUrl = `/apps/edit/${namespace}/${namespace}-${releaseName}/${e.currentTarget.value}`;
```
Comment (sameersbn) [inline dashboard/src/components/DeploymentForm/DeploymentForm.tsx]: triggers a version change

File: dashboard/src/containers/Root.tsx [hunk 9]
```diff
@@ -30,6 +31,8 @@ class Root extends React.Component {
   } = {
     "/": AppList,
     "/apps/:namespace/:releaseName": AppView,
+    "/apps/edit/:namespace/:releaseName": AppEdit,
+    "/apps/edit/:namespace/:releaseName/:version": AppEdit,
```
Comment (sameersbn) [inline dashboard/src/containers/Root.tsx]: additional route to specify a new version for a release

File: dashboard/src/components/DeploymentForm/DeploymentForm.tsx (hunk 3 above)
Comment (sameersbn) [reply dashboard/src/components/DeploymentForm/DeploymentForm.tsx]: update: version selection is now enabled

File: dashboard/src/components/AppEdit/AppEdit.tsx (hunk 6 above)
Comment (prydonius) [reply dashboard/src/components/AppEdit/AppEdit.tsx]: oops sorry, did not see you were passing all of this into DeploymentForm, ignore.

File: dashboard/src/components/DeploymentForm/DeploymentForm.tsx [hunk 10]
```diff
@@ -243,7 +256,10 @@ class DeploymentForm extends React.Component<IDeploymentFormProps, IDeploymentFo
     this.setState({ releaseName: e.currentTarget.value });
   };
   public handleChartVersionChange = (e: React.FormEvent<HTMLSelectElement>) => {
-    this.props.push(`/apps/new/${this.props.chartID}/versions/${e.currentTarget.value}`);
+    const { chartID, getChartVersion, getChartValues } = this.props;
+
+    getChartVersion(chartID, e.currentTarget.value);
```
Comment (prydonius) [inline dashboard/src/components/DeploymentForm/DeploymentForm.tsx]: we should only do this in AppEdit, if we do it in AppNew the version in the URL won't reflect what's in the state.

File: dashboard/src/containers/Root.tsx (hunk 9 above)
Comment (prydonius) [reply dashboard/src/containers/Root.tsx]: we can drop this now right?
